pennylane_lightning/core/src/bindings/Bindings.hpp

```diff
@@ -462,15 +462,20 @@ auto registerAdjointJacobian(
     using PrecisionT = typename StateVectorT::PrecisionT;
     std::vector<PrecisionT> jac(observables.size() * trainableParams.size(),
                                 PrecisionT{0.0});
+#ifdef _ENABLE_PLQUBIT
+    auto svdata = sv.getData();
+#elif _ENABLE_PLKOKKOS == 1
+    std::vector<Kokkos::complex<PrecisionT>> datavec(sv.getLength());
+    sv.DeviceToHost(datavec.data(), datavec.size());
+    auto svdata = datavec.data();
+#endif
```

Comment:
We can probably avoid the use of compiler preprocessor directives, by using unique pointers.
A simpler version of what we have [here](https://github.com/PennyLaneAI/pennylane-lightning/blob/master/pennylane_lightning/core/src/simulators/lightning_qubit/algorithms/AdjointJacobianLQubit.hpp#L253-L263) should suffice.

Reply:
I'm not sure that works because it requires branching depending on the state vector type, but the first condition may not be defined depending on the backend.

Reply:
Since the data type differs, using a u_ptr may not work here without some effort. We'd likely get this working with some combination of `if constexpr` but in the interest of time let's accept this as is.

Reply:
OK.